algebra2/basic_algebra_project/basic_algebra_2_pack.hpp

```diff
@@ -0,0 +1,670 @@
+}
+
+template <typename T, typename LP>
+inline auto algebra::kernel_2(ranges::matrix_view<T, LP> coefficients)
+    -> matrix<to_fraction_type<T>, LP> {
+  using namespace algorithms::gaussian_elimination;
+  auto t_coeff = ranges::transpose(coefficients);
+  auto [identity_vector, identity_matrix] =
+      eye<T, LP>(t_coeff.number_of_rows(), t_coeff.number_of_rows());
+  auto t_identity = ranges::transpose(identity_matrix);
+  auto [combined_vector, combined_matrix] =
+      combine_matrices(t_coeff, t_identity);
+  auto [solution_vector, solution_matrix] =
+      run(combined_matrix, reducted_form::diagonal);
+  return std::pair{std::move(solution_vector), solution_matrix};
+}
+
+template <typename T, typename LP>
+inline auto algebra::kernel_3(ranges::matrix_view<T, LP> matrix, int rows,
+                              int columns, int &l) -> std::vector<T> {
+  std::vector<T> result;
+  for (int i = 0; i < rows; i++) {
+    bool helping_variable = false;
+    for (int j = 0; j < columns; j++) {
+      if (matrix[i, j].numerator != 0) {
+        helping_variable = true;
+      }
+    }
+    if (!helping_variable) {
+      for (std::size_t j = columns; j < matrix.number_of_columns(); j++) {
+        result.push_back(matrix[i, j]);
+      }
+      l++;
+    }
+  }
+  return result;
+}
+
+template <typename T, typename LP>
+inline auto algebra::kernel(ranges::matrix_view<T, LP> coefficients)
+    -> std::pair<std::vector<to_fraction_type<T>>,
+                 ::ranges::matrix_view<to_fraction_type<T>, std::layout_left>> {
+  using namespace algorithms::gaussian_elimination;
+  int rows = coefficients.number_of_columns();
+  int columns = coefficients.number_of_rows();
+  auto [solution_vector, solution_matrix] = kernel_2(coefficients);
+  int result_rows = 0;
+  auto result = kernel_3(solution_matrix, rows, columns, result_rows);
+  if (result.size() != 0) {
+    ranges::matrix_view result_m(result, result_rows,
+                                 result.size() / result_rows, layout::column);
+    return std::pair{std::move(result), result_m};
+  }
+  result.push_back(to_fraction_type<T>{0});
+  ranges::matrix_view result_m(result, 1, 1, layout::column);
+  return std::pair{std::move(result), result_m};
+}
+
+template <typename T, typename LP>
+inline auto algebra::is_in_span(std::ranges::range auto v,
+                                ranges::matrix_view<T, LP> span)
+    -> algebra::is_in_span_result<to_fraction_type<T>> {
+  algebra::is_in_span_result<to_fraction_type<T>> result;
+  auto result_solve = algebra::solve(span, v);
+  if (!result_solve.exists) {
+    return result;
+  }
+  if (result_solve.special) {
+    result.belongs_to_span = true;
+    result.coefficients = result_solve.special.value();
+  }
+  return result;
+}
+
+template <typename T, typename LP>
+inline auto algebra::forms_base(ranges::matrix_view<T, LP> vectors) -> bool {
+  int rows = vectors.number_of_rows();
+  int columns = vectors.number_of_columns();
```

Comment:
no need to define new variables if you use them only once (by the way - you definitely do not pass clang tidy checks because these vars should be made const